aiida/backends/control.py

```diff
@@ -24,52 +22,71 @@
 
 MAINTAIN_LOGGER = AIIDA_LOGGER.getChild('maintain')
 
+# Note: these functions are not methods of `AbstractRepositoryBackend` because they need access to the orm.
+# This is because they have to go through all the nodes to gather the list of keys that AiiDA is keeping
+# track of (since they are descentralized in each node entry).
+# See the get_unreferenced_keyset function
+
 
 def repository_maintain(
+    manager: Manager,
     full: bool = False,
     dry_run: bool = False,
-    backend: Optional['Backend'] = None,
     **kwargs,
 ) -> None:
     """Performs maintenance tasks on the repository.
 
-        :param full:
-            flag to perform operations that require to stop using the maintained profile.
+    If `full == True`, then this method will attempt to block the profile to guarantee the
+    safety of its procedures. This will not only prevent any other subsequent process from
+    accessing that profile, but will also first check if there is already any process using
+    it and raise if that is the case. You will have to manually stop any processes that is
+    currently accessing the profile yourself or wait for them to finish on their own.
 
-        :param dry_run:
-            flag to only print the actions that would be taken without actually executing them.
+    :param manager:
+        a Manager instance to control the profile/backend to be maintained.
 
-        :param backend:
-            specific backend in which to apply the maintenance (defaults to current profile).
-    """
+    :param full:
+        flag to perform operations that require to stop using the profile/backend to be maintained.
 
-    if backend is None:
-        backend = get_manager().get_backend()
+    :param dry_run:
+        flag to only print the actions that would be taken without actually executing them.
+
+    """
+    type_check(manager, Manager)
+    profile = manager.get_profile()
+    backend = manager.get_backend()
     repository = backend.get_repository()
 
-    unreferenced_objects = get_unreferenced_keyset(aiida_backend=backend)
-    MAINTAIN_LOGGER.info(f'Deleting {len(unreferenced_objects)} unreferenced objects ...')
-    if not dry_run:
-        repository.delete_objects(list(unreferenced_objects))
+    def perform_tasks():
+        unreferenced_objects = get_unreferenced_keyset(aiida_backend=backend)
+        MAINTAIN_LOGGER.info(f'Deleting {len(unreferenced_objects)} unreferenced objects ...')
+        if not dry_run:
+            repository.delete_objects(list(unreferenced_objects))
+
+        MAINTAIN_LOGGER.info('Starting repository-specific operations ...')
+        repository.maintain(live=not full, dry_run=dry_run, **kwargs)
 
-    MAINTAIN_LOGGER.info('Starting repository-specific operations ...')
-    repository.maintain(live=not full, dry_run=dry_run, **kwargs)
+    if full:
+        with ProfileAccessManager(profile).lock():
+            perform_tasks()
+    else:
+        perform_tasks()
```

Comment:
For this kind of paradigm where the contextmanager is conditional or optional, I personally prefer the following approach:
```suggestion
    from contextlib import nullcontext
    
    if full:
        context = ProfileAccessManager(profile).lock
    else:
        context = nullcontext
        
    with context():
        unreferenced_objects = get_unreferenced_keyset(aiida_backend=backend)
        MAINTAIN_LOGGER.info(f'Deleting {len(unreferenced_objects)} unreferenced objects ...')
        if not dry_run:
            repository.delete_objects(list(unreferenced_objects))

        MAINTAIN_LOGGER.info('Starting repository-specific operations ...')
        repository.maintain(live=not full, dry_run=dry_run, **kwargs)
```

Reply:
Uhm, I didn't know this `nullcontext` option. That is neat, thanks!